### PR TITLE
fix: Divide by zero should fail to satisfy constraints for `Field` and ints

### LINKED ISSUE #2149: Divide-by-zero is a constraint failure for integers but not for `Field`
### Aim

The proving phase would complain that constraints are unsatisfiable when dividing an unsigned integer by zero:

```rust
let a: u8 = 3 / 0;
```

However, when changing the variable type to Field, the proving phase would not complain:

```rust
let a: Field = 3 / 0;
```

It would be better to make the behavior consistent. Also, the compiler should explicitly detect and give more informative error messages for divide-by-zero.

### Expected Behavior

The compiler should explicitly issue an error message for divide-by-zero.

### Bug

Please see steps to reproduce.

### To Reproduce

Save the following code into `src/main.nr`:

```rust
use dep::std;

fn main() {
    let a: u8 = 3 / 0;
    std::println(a);
}
```

Run the following command:

```bash
nargo prove
```

You would see the following error message:

```bash
error: Unsatisfied constraint
  ┌─ /home/Jurarpereurs/src/main.nr:4:17
  │
4 │     let a: u8 = 3 / 0;
  │                 ----- Constraint failed

Error: could not satisfy all constraints

Location:
    crates/nargo_cli/src/cli/mod.rs:74:5
```

Then change the variable type to Field:

```rust
use dep::std;

fn main() {
    let a: Field = 3 / 0;
    std::println(a);
}
```

Run the following command:

```bash
nargo prove
```

The proving phase would succeed and you would see the following output:

```bash
"0x00"
```

### Installation Method

Binary

### Nargo Version

nargo 0.9.0 (git version hash: 35404ba9b2916cebf35519546eec0f0ae54b5516, is dirty: false)

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by vezenovm**
# Description

## Problem\*

Resolves #2149 

## Summary\*

The issue was posted using nargo 0.9.0 and there has since been some changes, although the bug still exists. `let a: Field = 3 / 0;` still unexpectedly prints zero while `let a: u8 = 3 / 0;` actually now panics during instruction simplification for attempting to divide by zero. `let a: i8 = 3 / 0;` fails in the same cases as `u8`. 

When the dividend comes from inputs to main, `Field` behaves the same way, while `u8` panics inside of execution of the Brillig VM. The Brillig VM should return an error for division by zero rather than panic'ing but this is a separate issue. 

The following changes were made:
 - For a field division, we have a check in ACIR gen for whether the brillig inverse is valid. However, we were missing this check for constants, this has now been included. 
- I have also introduced constraints to check whether we are dividing by zero in euclidean division. If we are dividing by zero, the execution of the program will fail to satisfy the constraints. 
- I had to remove a simplification for when we have an rhs of zero with a division binary op. Now, instead of a panic on `let a: u8 = 3 / 0` we will get a constraint failure that points to the span `3 / 0`. 

All variants of division by zero (whether by constant or witness) now trigger a failed constraint. 

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: crates/noirc_frontend/src/parser/parser.rs [hunk 1]
```diff
@@ -549,18 +547,16 @@ fn where_clause() -> impl NoirParser<Vec<TraitConstraint>> {
...
             result
         })
 }
 
 fn trait_bounds() -> impl NoirParser<Vec<TraitBound>> {
-    trait_bound()
-        .separated_by(just(Token::Plus))
-        .at_least(1)
-        .allow_trailing()
+    trait_bound().separated_by(just(Token::Plus)).at_least(1).allow_trailing()
```
Comment (vezenovm) [inline crates/noirc_frontend/src/parser/parser.rs:559]: This is cargo fmt

File: crates/nargo_cli/tests/compile_failure/divide_by_zero/src/main.nr [hunk 2]
```diff
@@ -0,0 +1,6 @@
+use dep::std;
+
+fn main() {
```
Comment (kevaundray) [inline crates/nargo_cli/tests/compile_failure/divide_by_zero/src/main.nr:3]: can you add a test case where the numerator is not a constant, so a parameter to main, and then one where both the numerator and denominator are non constant. This should fail execution

Comment (vezenovm) [reply crates/nargo_cli/tests/compile_failure/divide_by_zero/src/main.nr:3]: sure thing

**Review by kevaundray — APPROVED** [review]

**Review by TomAFrench — CHANGES_REQUESTED** [review]
Blocking this temporarily as this is a breaking change

**Comment by kevaundray** [comment]
> Blocking this temporarily as this is a breaking change

I don't think anyone was relying on their program panicking before

**Comment by vezenovm** [comment]
> I don't think anyone was relying on their program panicking before

Or a Field division being `0`. If anything this will give them the correct result now no?

**Review by TomAFrench — APPROVED** [review]

**Pull request merged**